packages/api/src/lib/granule-delete.ts

```diff
@@ -0,0 +1,100 @@
+import Knex from 'knex';
+import pMap from 'p-map';
+
+import { deleteS3Object } from '@cumulus/aws-client/S3';
+import {
+  GranulePgModel,
+  FilePgModel,
+  PostgresGranuleRecord,
+  PostgresFileRecord,
+} from '@cumulus/db';
+import { DeletePublishedGranule } from '@cumulus/errors';
+import { ApiFile, ApiGranule } from '@cumulus/types';
+
+const FileUtils = require('../../lib/FileUtils');
+const Granule = require('../../models/granules');
+
+/**
+ * Delete a list of files from S3
+ *
+ * @param {Array} files - A list of S3 files
+ * @returns {Promise}
+ */
+const _deleteS3Files = async (
+  files: (ApiFile | PostgresFileRecord)[] = []
+) =>
+  pMap(
+    files,
+    (file) => {
+      deleteS3Object(
+        FileUtils.getBucket(file),
+        FileUtils.getKey(file)
+      );
+    }
+  );
+
+/**
+ * Delete a Granule from Postgres and Dynamo, delete the Granule's
+ * Files from Postgres and S3
+ *
+ * @param {Object} params
+ * @param {Knex} params.knex - DB client
+ * @param {Object} params.dynamoGranule - Granule from DynamoDB
+ * @param {PostgresGranule} params.pgGranule - Granule from Postgres
+ * @param {FilePgModel} params.filePgModel - File Postgres model
+ * @param {GranulePgModel} params.granulePgModel - Granule Postgres model
+ * @param {Object} params.granuleModelClient - Granule Dynamo model
+ */
+const deleteGranuleAndFiles = async ({
+  knex,
+  dynamoGranule,
+  pgGranule,
+  filePgModel = new FilePgModel(),
+  granulePgModel = new GranulePgModel(),
+  granuleModelClient = new Granule(),
+}: {
+  knex: Knex,
+  dynamoGranule: ApiGranule,
+  pgGranule: PostgresGranuleRecord,
+  filePgModel: FilePgModel,
+  granulePgModel: GranulePgModel,
+  granuleModelClient: typeof Granule
+}) => {
+  if (pgGranule === undefined) {
+    // Delete only the Dynamo Granule and S3 Files
+    await _deleteS3Files(dynamoGranule.files);
+    await granuleModelClient.delete(dynamoGranule);
+  } else if (pgGranule.published) {
+    throw new DeletePublishedGranule('You cannot delete a granule that is published to CMR. Remove it from CMR first');
+  } else {
+    // Delete PG Granule, PG Files, Dynamo Granule, S3 Files
+    const files = await filePgModel.search(
+      knex,
+      { granule_cumulus_id: pgGranule.cumulus_id }
+    );
+
+    await knex.transaction(async (trx) => {
+      await pMap(
+        files,
+        (file) => {
+          filePgModel.delete(trx, { cumulus_id: file.cumulus_id });
+        }
+      );
+
+      // TODO: relying on the cumulus_id from the lookup is icky, but we need to
+      // truly identify the unique record.
+      await granulePgModel.delete(trx, {
+        granule_id: pgGranule.granule_id,
+        collection_cumulus_id: pgGranule.collection_cumulus_id,
+        // cumulus_id: pgGranule.cumulus_id,
+      });
+      await granuleModelClient.delete(dynamoGranule);
+    });
+
+    await _deleteS3Files(files);
```

Comment:
Should deleting the s3 files be in the transaction as well? Also, is there a significance to deleting the s3 files before deleting the dynamo granule?

Reply:
As part of the original PR discussion for this in the granule delete ticket we went with  

1) data delete
2) S3 delete

As a situation where there's orphaned files on the FS is  preferable to orphaned DB records -- if any part of the DB update fails, the transaction reverts and the S3 objects are still on the FS.   If we move the file deletions in, any file failing will result in a full rollback of the transaction.     But I suppose we could (and probably should) hold the transaction (and row locks) until the files are gone.... except that doesn't actually buy us anything except FS bound lock times *because*:

*updates* to the file system are not concurrent with the database updates related to moving those files.      Duplicate granule scenarios could be fraught, particularly if the overwrite winds up being a different file because the writes only check the state of S3.   

🤔 

This bears some discussion in the AM... but it's the same logic from granule-delete.js

Reply:
Yeah, this is just a direct copy of the logic from `granule-delete.js` into TS. If we want to revisit the logic, that's fine, but it's a bit outside the scope of this PR/work